src/building/bootstrapping.md

```diff
@@ -47,54 +62,59 @@ graph TD
+them to compile from current code a compiler. This is the stage 1 compiler.
 
-The rustc source code is then compiled with the stage0 compiler to produce the stage1 compiler.
+The stage 1 compiler is the first that is from current code. Yet, it is not
+entirely up-to-date, because the compiler that compiled it is of earlier code.
+More on this below.
 
-### Stage 2
+#### Stage 2: the truly current compiler
 
-We then rebuild our stage1 compiler with itself to produce the stage2 compiler.
+By default, the stage 1 libraries are copied into stage 2, because they are
+expected to be identical.
 
-In theory, the stage1 compiler is functionally identical to the stage2 compiler,
-but in practice there are subtle differences.
-In particular, the stage1 compiler itself was built by stage0
-and hence not by the source in your working directory.
-This means that the symbol names used in the compiler source
-may not match the symbol names that would have been made by the stage1 compiler,
-which can cause problems for dynamic libraries and tests.
+The stage 1 compiler is used to compile from current code a compiler. This is
+the stage 2 compiler.
 
-The `stage2` compiler is the one distributed with `rustup` and all other install methods.
-However, it takes a very long time to build
-because one must first build the new compiler with an older compiler
-and then use that to build the new compiler with itself.
-For development, you usually only want the `stage1` compiler,
-which you can build with `./x.py build library`.
+The stage 2 compiler is the first that is both from current code and compiled
+by a compiler that is of current code. The compilers and libraries obtained by
+`rustup` and other installation methods are all stage 2.
+
+For most purposes a stage 1 compiler would suffice: `x.py build library`.
 See [Building the Compiler](./how-to-build-and-run.html#building-the-compiler).
+Between the stage 2 and the stage 1 compiler are subtle differences:
+
+The symbol names used in the compiler source may not match the symbol names
+that would have been made by the stage1 compiler. This is important when using
+dynamic linking and due to the lack of ABI compatibility between versions. This
+primarily manifests when tests try to link with any of the `rustc_*` crates or
+use the (now deprecated) plugin infrastructure. These tests are marked with
+`ignore-stage1`.
+
+Also, the stage 2 compiler benefits from the compile-time optimizations
+generated by a compiler that is of the current code.
 
-### Stage 3
+#### Stage 3: the same-result test
 
-Stage 3 is optional. To sanity check our new compiler, we
-can build the libraries with the stage2 compiler. The result ought
-to be identical to before, unless something has broken.
+To verify that the stage 2 libraries that were copied from stage 1 are indeed
+identical to those which would otherwise have been compiled in stage 2, the
+stage 2 compiler is used to compile them and a comparison is made.
```

Comment:
I guess emphasizing that stage 3 is optional is important here as most contributors don't need it.

Reply:
What is it, more specifically, that would require one to compile or use this stage compiler?

Reply:
Please determine whether it is resolved in HEAD.